### PR TITLE
Enforce buffer size limits for BlockEncodingBuffer

### LINKED ISSUE #14162: OptimizedPartitionedOutputOperator uses too much memory
Recently we observed OptimizedPartitionedOutputOperator is using more memory than the legacy PartitionedOutputOperator. There're a few facts that contributed to this increased memory usage:

1. OptimizedPartitionedOutputOperator persists one or several BlockEncodingBuffers for each column for each partition. These buffers won't shrink during the query life time.
2. For nested  BlockEncodingBuffer, it uses extra int[] to hold the positions, mapped positions and offsets. If the array/map cells are too large (e.g. over a few million entries per row), this positions array itself takes a lot of space.
3. Each BlockEncodingBuffer maintains several byte[] buffers to hold the data. These buffers are created when the first page comes in, and are being grown by a factor of 2 on the fly. This may cause the buffer to grow larger than necessary. Reducing the growth factor to 1.5 regressed the performance, therefore we want to keep the growth factor 2 but enforce stricter limit of the buffer sizes.
4. We add at least one top level row into the buffer at a time. Some cell may be very big (e.g. array with millions of entries), and the buffer for this single row would be up to several dozens of MB. In this case, the size limit "driver.max-page-partitioning-buffer-size" would be ignored.

We propose the following changes to address the problem:

1. Reuse the arrays

The OptimizedPartitionedOutputOperator adds the data from incoming page to buffers partition by partition, and within each partition column by column. There's opportunity we can reuse the buffers for different partitions. In the following PRs, we reused the int[] and byte[] buffers by using SimpleArrayAllocator:

- https://github.com/prestodb/presto/pull/14234 : Reuse buffers in OptimizedPartitionedOutputOperator - part 1 serializedRowSizes, offsetsCopy, positions, mappedPositions and offsets
- https://github.com/prestodb/presto/pull/14286 : Reuse buffers in OptimizedPartitionedOutputOperator - part 2 : byte[] buffers in BlockEncodingBuffer

2. Reduce the positions/mappedPositions array sizes

Previously the positions array was initialized to be the incoming page's positionCount size. This could waste lots of memory since each partition may only get a small portion of rows. https://github.com/prestodb/presto/pull/14222 reduces the initial positions array size from positionCount to `min(positionCount, (positionCount / partitionCount + 1) * 2)` and grow it on the fly

3. Enforce buffer size limits for BlockEncodingBuffer
https://github.com/prestodb/presto/pull/14295 uses incoming blocks' retained sizes as an estimation for how large the buffers can grow. Before this limit was reached, the byte[] buffers were grown with growth factor 2. Once this limit is reached the growth factor would become 1.

4. Bug fixes in memory counting
- https://github.com/prestodb/presto/pull/14297 Fix getRetainedSizeInBytes for DecodedBlockNode
- https://github.com/prestodb/presto/pull/14296 Fix getRetainedSizeInBytes for ColumnarXXX
- https://github.com/prestodb/presto/pull/14249 Fix memory tracking for PartitionedOutputOperator

**Further work:**

https://github.com/prestodb/presto/pull/14234 and https://github.com/prestodb/presto/pull/14286 reuses the buffers between partitions. Ie. these buffers can be reused by the next partition if the current page can be consumed totally for the current partition. This can be further improved to be reused between different BlockEncodingBuffers within the same partition, if we can move the calculation of row sizes into the Page/Block interface. Currently we need to calculate the row sizes in BlockEncodingBuffers after the positions/offsets are populated for all BlockEncodingBuffers in the same partition. This forbids reusing the buffers among BlockEncodingBuffers within the same partition. If we can move the calculation of row sizes out of BlockEncodingBuffer, then we can populate the positions/offsets block by block within each partition
```
for each incoming page, 
    populate top level positions for each partition;
    calculate rowsizes by calling page.getSerializedRowSizes(int[] serializedRowSizes);
for each partition
    while the current page is not totally consumed
        calculate batchSize for current batch;
        for each column i
            populate nested level positions, mappedpositions and offsets for the whole block if they are not populated yet(first batch in page)
            blockEncodingBuffers[i].appendDataInBatch();
            flush the current buffer if full.
            if the last batch, return nested level positions, mappedpositions and offsets to allocator
            if the last batch causes a flush, return the byte[] buffers to allocator
```
Most of times, a page can be consumed in a single batch. This new work flow would allow these buffers to be reused among blockEncodingBuffers as well as among partitions. However we may need to add page#getSerializedRowSizes(int[] serializedRowSizes) to Page and Block interface.

### PR DESCRIPTION
**Pull request by yingsu00**
Partially resolves #14162
Depend on https://github.com/prestodb/presto/pull/14286

The byte[] buffers in BlockEncodingBuffer used to be grown by a factor
of 2. This could cause the buffers larger than necessary. Reducing
the expansionFactor has negative impact on performance. Therefore we
want to enforce stricter limit when growing these buffers. This PR
uses incoming blocks' retained sizes as an estimation for how large the
buffers can grow. It shows 30% to 40% memory usage reduction for fixed
width types and 10% reduction for variable width types for the
OptimizedPartitionedOutputOperator.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-array/src/main/java/com/facebook/presto/array/Arrays.java [hunk 1]
```diff
@@ -104,9 +106,9 @@ else if (buffer.length < capacity) {
         return buffer;
     }
 
-    public static byte[] ensureCapacity(byte[] buffer, int capacity, ExpansionFactor expansionFactor, ExpansionOption expansionOption, ArrayAllocator allocator)
+    public static byte[] ensureCapacity(byte[] buffer, int capacity, int targetSize, ExpansionFactor expansionFactor, ExpansionOption expansionOption, ArrayAllocator allocator)
```
Comment (mbasmanova) [inline presto-array/src/main/java/com/facebook/presto/array/Arrays.java]: targetSize -> estimatedMaxCapacity for consistency

File: presto-array/src/main/java/com/facebook/presto/array/Arrays.java [hunk 2]
```diff
@@ -115,7 +117,7 @@ else if (buffer.length < capacity) {
         else if (buffer.length < capacity) {
             newBuffer = allocator.borrowByteArray(newCapacity);
             if (expansionOption == PRESERVE) {
-                System.arraycopy(buffer, 0, newBuffer, 0, Math.min(buffer.length, capacity));
+                System.arraycopy(buffer, 0, newBuffer, 0, min(buffer.length, capacity));
```
Comment (mbasmanova) [inline presto-array/src/main/java/com/facebook/presto/array/Arrays.java]: min(buffer.length, capacity) -> buffer.length

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java [hunk 3]
```diff
@@ -201,17 +204,19 @@ public String toString()
...
         decodedBlockNode = mapPositionsToNestedBlock(decodedBlockNode);
         ColumnarArray columnarArray = (ColumnarArray) decodedBlockNode.getDecodedBlock();
         decodedBlock = columnarArray.getNullCheckBlock();
 
+        setTargetBufferSize(decodedBlockNode, partitionBufferCapacity);
+
         populateNestedPositions(columnarArray);
 
-        ((AbstractBlockEncodingBuffer) valuesBuffers).setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+        ((AbstractBlockEncodingBuffer) valuesBuffers).setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0), partitionBufferCapacity);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java]: Instead of casting in multiple places change the type of valuesBuffers to AbstractBlockEncodingBuffer

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java [hunk 4]
```diff
@@ -201,17 +204,19 @@ public String toString()
...
-    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
+    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode, int partitionBufferCapacity)
     {
         requireNonNull(decodedBlockNode, "decodedBlockNode is null");
 
         decodedBlockNode = mapPositionsToNestedBlock(decodedBlockNode);
         ColumnarArray columnarArray = (ColumnarArray) decodedBlockNode.getDecodedBlock();
         decodedBlock = columnarArray.getNullCheckBlock();
 
+        setTargetBufferSize(decodedBlockNode, partitionBufferCapacity);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java]: I suggest to inline setTargetBufferSize. It is used only once and doesn't actually *set* target size.

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/DecodedBlockNode.java [hunk 5]
```diff
@@ -42,10 +43,16 @@
     private final Object decodedBlock;
     private final List<DecodedBlockNode> children;
 
-    public DecodedBlockNode(Object decodedBlock, List<DecodedBlockNode> children)
+    // The portion of the size of the exclusive content of this block to the total size. For example, the exclusive content
+    // of an ArrayBlock is its offsets and nulls array while not its elementBlock. It is used to estimate the limit of the
+    // buffers' sizes.
+    private double sizePortionInPage;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/DecodedBlockNode.java]: Can this be calculated within this class as getRetainedSizeInBytes - sum(child.getRetainedSizeInBytes())? If not, perhaps, move this change into a separate commit.

File: presto-array/src/main/java/com/facebook/presto/array/Arrays.java (hunk 1 above)
Comment (yingsu00) [reply presto-array/src/main/java/com/facebook/presto/array/Arrays.java]: Done. I renamed the BlockEncodingBuffer fields names too. E.g. `valuesBufferTargetSize` --> `estimatedValueBufferMaxCapacity`

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java (hunk 3 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java]: Done in new commit `17ee832a6a Change nested buffers to AbstractBlockEncodingBuffer`

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/DecodedBlockNode.java (hunk 5 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/DecodedBlockNode.java]: @mbasmanova Yes you're actually right. I moved the logic to DecodedBlockNode and made a new commit `52456a0fec Introduce sizePortionInPage in DecodedBlockNode`

**Comment by yingsu00** [comment]
@mbasmanova Thanks for reviewing! I have addressed all your comments. Will you be able to take another look? Thank you!

**Comment by mbasmanova** [comment]
@yingsu00 Thanks. This PR shows conflicts with master. Please, rebase and I'll take another pass.

**Comment by yingsu00** [comment]
> @yingsu00 Thanks. This PR shows conflicts with master. Please, rebase and I'll take another pass.

@mbasmanova Hi Masha, I just rebased. Would you please take a look please?

**Comment by yingsu00** [comment]
@mbasmanova Hi Masha, the motivation is to get an estimate of the buffers' capacity as close as possible. The estimation is based on the retained sizes of the blocks in the current page. For example, a page with two blocks has the following retained sizes:

channel 0: LongArrayBlock(block0) of 100KB
channel 1: ArrayBlock(block1) of 900KB with nested LongArrayBlock(block2) of 800KB
All blocks don't have nulls. Then block0's portion in the page is `100KB/(100KB + 900KB) = 10%`, while block1 's portion (for the offsets) is `(900KB-800KB)/(100KB + 900KB) = 10%`, and block2's portion is `800KB/(100KB + 900KB) = 80%`.

Now suppose the PartitionBuffer's limit is 1MB, then block0's valuesBuffer max capacity estimation is 0.1MB, block1's offsetsBuffer max capacity estimation is 0.1MB, and block2's valuesBuffer max capacity estimation is 0.8MB. If the buffers grow beyond this limit, the expansion factor will become 1.

**Comment by yingsu00** [comment]
@mbasmanova Hi Masha, did my previous reply answer your question? Would adding more comments to the code help? Or do you have other ideas to address the problem? This PR reduces memory consumption by 30% for a lot of cases without impacting the performance too much and I'd love to hear your suggestion and get it merged. Thank you!

**Comment by mbasmanova** [comment]
@yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?

Aside from the question above, I think the code would be simpler if this calculation was localized. Specifically, I suggest the following changes.

- DecodedBlock: Replace sizePortionInPage with childrenRetainedSizeInBytes 

```
        long childrenSize = 0;
        for (int i = 0; i < children.size(); i++) {
            childrenSize += children.get(i).getRetainedSizeInBytes();
        }

        childrenRetainedSizeInBytes = childrenSize;
```

- PagePartitioner: Add pageRetainedSizeInBytes parameter to appendData method and pass this value to setupDecodedBlocksAndPositions

```
            long pageRetainedSizeInBytes = page.getRetainedSizeInBytes();
            for (int i = 0; i < decodedBlocks.length; i++) {
                decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i), pageRetainedSizeInBytes);
            }

            // Copy the data to their destination partitions and flush when the buffer is full.
            for (int i = 0; i < partitionBuffers.length; i++) {
                partitionBuffers[i].appendData(decodedBlocks, pageRetainedSizeInBytes, fixedWidthRowSize, variableWidthChannels, outputBuffer);
            }
...
            for (int i = 0; i < channelCount; i++) {
                blockEncodingBuffers[i].setupDecodedBlocksAndPositions(decodedBlocks[i], positions, positionCount, capacity, pageRetainedSizeInBytes);
            }
```

- Calculate block's fraction of the input page size in setupDecodedBlocksAndPositions and pass it to setupDecodedBlockAndMapPositions (decodedBlockPageSizeFraction below needs a better name, but I can't thing of one just yet):

```
    @Override
    public void setupDecodedBlocksAndPositions(DecodedBlockNode decodedBlockNode, int[] positions, int positionCount, int partitionBufferCapacity, long pageRetainedSizeInBytes)
    {
        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
        requireNonNull(positions, "positions is null");

        this.positions = positions;
        this.positionCount = positionCount;
        this.positionsOffset = 0;
        this.positionsMapped = false;

        double decodedBlockPageSizeFraction = (decodedBlockNode.getRetainedSizeInBytes() - decodedBlockNode.getChildrenRetainedSizeInBytes()) / ((double) pageRetainedSizeInBytes);

        setupDecodedBlockAndMapPositions(decodedBlockNode, partitionBufferCapacity, decodedBlockPageSizeFraction);
    }
```

- Replace calls to DecodedBlock.getSizePortionInPage() with decodedBlockPageSizeFraction

**Comment by yingsu00** [comment]
> @yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?

@mbasmanova Hi Masha, yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is `positionCount * averageRowSize of dictionary`. What do you think about it? Just to mention that this won't cause functional issues but may cause the buffers to be grown slower.

And thanks for the suggestion of moving the logic into `setupDecodedBlocksAndPositions`. Actually I tried to do as much arithmetic calculation as possible just once for each incoming page. By moving the logic inside of the BlockEncodingBuffer, we're repeating the calculation of minus and division for all partitions/BlockEncodingBuffers which is unnecessary. If we do the minus and division in DecodedBlock, then we just need to do a multiplication in each partitions/BlockEncodingBuffers. Originally I was thinking to calculate the buffer size in DecodedBlock but they are orthogonal and may not be a good idea. Are you strongly against the current way? If yes I can do a JMH run to see if it has any regressions.

**Comment by yingsu00** [comment]
> > @yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?
> 
> @mbasmanova Hi Masha, yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is `positionCount * averageRowSize of dictionary`. What do you think about it? Just to mention that this won't cause functional issues but may cause the buffers to be grown slower.
> 
> And thanks for the suggestion of moving the logic into `setupDecodedBlocksAndPositions`. Actually I tried to do as much arithmetic calculation as possible just once for each incoming page. By moving the logic inside of the BlockEncodingBuffer, we're repeating the calculation of minus and division for all partitions/BlockEncodingBuffers which is unnecessary. If we do the minus and division in DecodedBlock, then we just need to do a multiplication in each partitions/BlockEncodingBuffers. Originally I was thinking to calculate the buffer size in DecodedBlock but they are orthogonal and may not be a good idea. Are you strongly against the current way? If yes I can do a JMH run to see if it has any regressions.

@mbasmanova  Hi Masha, I changed the code as you suggested and just pushed the PR. No regression was observed. Thank you for the detailed write-out! Will you take another look?

**Comment by mbasmanova** [comment]
@yingsu00 I think `Introduce sizePortionInPage in DecodedBlockNode` commit needs to be removed.

> yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is positionCount * averageRowSize of dictionary. What do you think about it? 

Sounds reasonable.

**Comment by yingsu00** [comment]
> positionCount * averageRowSize
@mbasmanova Hi Masha, I removed `Introduce sizePortionInPage in DecodedBlockNode` commit and added a new commit `Use inflated size for dictionary and RLE blocks in DecodedBlockNode` to show you the change for DictionaryBlock and RunLengthEncodedBlock retained size estimation. But please also note that for Block views, this estimation is still inaccurate. The retained size is the size for whole arrays but the positionCount for the raw ids is not accessible. If we want to solve this problem, we could maintain both sizeInBytes and retainedSizeInBytes for DecodedBlockNode, and use sizeInBytes to estimate the buffer growth limit since it reflects only the size of the region for the views. What do you think?

**Comment by mbasmanova** [comment]
@yingsu00 

>  But please also note that for Block views, this estimation is still inaccurate.

What is "Block view"?

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 6]
```diff
@@ -479,13 +480,15 @@ public void partitionPage(Page page)
...
+            decodedBlocksRetainedSize = 0;
             for (int i = 0; i < decodedBlocks.length; i++) {
                 decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i));
+                decodedBlocksRetainedSize += decodedBlocks[i].getRetainedSizeInBytes();
             }
 
             // Copy the data to their destination partitions and flush when the buffer is full.
             for (int i = 0; i < partitionBuffers.length; i++) {
-                partitionBuffers[i].appendData(decodedBlocks, page.getRetainedSizeInBytes(), fixedWidthRowSize, variableWidthChannels, outputBuffer);
+                partitionBuffers[i].appendData(decodedBlocks, decodedBlocksRetainedSize, fixedWidthRowSize, variableWidthChannels, outputBuffer);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:491]: @yingsu00 I think moving away from page.getRetainedSizeInBytes() and using estimated flattened size is the right direction. It allows for much cleaner design. The DecodedBlockNode still needs the original value for the retainedSizeInBytes as it is used in memory tracking. Hence, I suggest to introduce a new variable to store the estimate of the flattened size of the block.

**Comment by yingsu00** [comment]
> @yingsu00
> 
> > But please also note that for Block views, this estimation is still inaccurate.
> 
> What is "Block view"?

By "Block view" I meant the block with non-zero offsetBase.

**Comment by yingsu00** [comment]
@mbasmanova Hi Masha, I found the Block interface has getLogicalSizeInBytes() and it can be used to estimate the serialized sizes. For DictionaryBlock and RunLengthEncodedBlock, it is the size after the block is expanded. For other blocks is the sizeInBytes which just account for the region that is going to be serialized. I think it's pretty accurate for our purpose. I updated the PR with two new commits
```
Introduce estimatedSerializedSizeInBytes in ColumnarXXX
Introduce estimatedSerializedSizeInBytes in DecodedBlockNode
```
I also modified `Enforce buffer size limits for BlockEncodingBuffer` to use the newly introduced `estimatedSerializedSizeInBytes`, and removed `Introduce sizePortionInPage in DecodedBlockNode`. The tests have already passed. Would appreciate your review again. Thank you!
